### DISCUSSION TITLE
form.Field children prop: Render props pattern with children as a function

### DISCUSSION THREAD

**Discussion by arnaudbzn** [TanStack/form#476] [Ideas] [answered]
The render props pattern offers great flexibility. 
But in the docs and examples using the `children` prop in the `Field` component is a bit unusual (though it makes sense for a `renderComponent` prop).

```TypeScript
     <form.Field children={(field) => ( ... )}/>
```

It's may be more intuitive to use this syntax for the children:

```TypeScript
    <form.Field>
       {(field) => ( ... )}
    </form.Field>
```

Remix uses this approach with the [Await](https://remix.run/docs/en/main/components/await#await) component

The feature and implementation are the same, it's just a suggestion for the documentation and examples.

**Comment by crutchcorn** [comment]
I broadly agree FWIW, and is how I did docs as part of [HouseForm](houseform.dev) (the previous form library I built).

**Comment by bzbetty** [comment]
would the following syntax be an option?

```
   {form.field({
       name: ''
       children: () => {}
   })
```

**Reply by crutchcorn** [reply]
While this syntax is something we'll likely be using for future adapters, it's not feasible to use this syntax for our React adapter thanks to performance concerns:

https://unicorn-utterances.com/posts/functions-are-killing-react-performance

**Comment by mmospanenko** [comment]
how to be with ESLint [react/no-children-prop](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-children-prop.md) rule?

**Reply by dca123** [reply, marked as answer]
> how to be with ESLint [react/no-children-prop](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-children-prop.md) rule?

You can add this to your eslintrc.

```
	"rules": {
		"react/no-children-prop": [
			true,
			{
				"allowFunctions": true
			}
		]
	}
```


Alternatively you can do

```
<form.Field >
   {(field) => (<></>) }
</form.Field>
```

**Reply by JPGarCar** [reply]
I might not have seen this on the docs, but if it isn't, then it should be. Thank you for the tip!

**Comment by martin-dimi** [comment]
Is the expectation to still ignore that ESlint rule? @crutchcorn

**Reply by LeCarbonator** [reply]
If you have the ESLint rule enabled, you can just use the other way to write children in React:
```tsx
<A>
  <B />
</A>

<A children={<B />} />

<A>
  {() => <C />}
</A>

<A children={() => <C />}/>
```

**Reply by luchillo17** [reply]
Why not use the children within tags by default in the docs then?

**Reply by LeCarbonator** [reply]
> Why not use the children within tags by default in the docs then?

Because it's a preference. Some ESLint configurations force you to use `children` as prop if it's a function and as html child otherwise. It's up to the user as well as the configuration they chose.

**Reply by luchillo17** [reply]
Yeah now that I think about it the default rule is not the only way, there are projects that will use the props way, still the wrapped children is the most common rule setup in ESLint so I still think the docs should use that.
